extension/background.js

```diff
@@ -0,0 +1,199 @@
+      ["blocking"]
+    );
+
+    this.debug("registering onRedirected listener");
+    browser.addonsSearchExperiment.onRedirected.addListener(
+      this.onRedirectedListener,
+      { urls: patterns }
+    );
+  }
+
+  // This listener is required to force the registration of traceable channels.
+  noOpListener() {
+    // Do nothing.
+  }
+
+  async onRedirectedListener({ addonId, firstUrl, lastUrl }) {
+    if (!firstUrl || !lastUrl) {
+      // Something went wrong but there is nothing we can do at this point.
+      return;
+    }
+
+    // When we do not have an add-on ID (in the request property bag), we
+    // likely detected a search server-side redirect.
+    const maybeServerSideRedirect = !addonId;
+
+    let addonIds = [];
+    // Search server-side redirects are possible because an extension has
+    // registered a search engine, which is why we can (hopefully) retrieve the
+    // add-on ID.
+    if (maybeServerSideRedirect) {
+      addonIds = this.getAddonIdsForUrl(firstUrl);
+    } else if (addonId) {
+      addonIds = [addonId];
+    }
+
+    if (addonIds.length === 0) {
+      // No add-on ID means there is nothing we can report.
+      return;
+    }
+
+    // This is the initial URL before any redirect.
+    const from = await browser.addonsSearchExperiment.getPublicSuffix(firstUrl);
+    // This is the final URL after redirect(s).
+    const to = await browser.addonsSearchExperiment.getPublicSuffix(lastUrl);
+
+    if (from === to) {
+      // We do not want to report redirects to same public suffixes. However,
+      // we will report redirects from public suffixes belonging to a same
+      // entity (.e.g., `example.com` -> `example.fr`).
+      //
+      // Known limitation: if a redirect chain starts and ends with the same
+      // public suffix, we won't report any event, even if the chain contains
+      // different public suffixes in between.
+      return;
+    }
+
+    const telemetryObject = maybeServerSideRedirect
+      ? TELEMETRY_OBJECT_OTHER
+      : TELEMETRY_OBJECT_WEBREQUEST;
+    const telemetryValue = maybeServerSideRedirect
+      ? TELEMETRY_VALUE_SERVER
+      : TELEMETRY_VALUE_EXTENSION;
+
+    for (const addonId of addonIds) {
+      const addonVersion = await browser.addonsSearchExperiment.getAddonVersion(
+        addonId
+      );
+      const extra = { addonId, addonVersion, from, to };
+
+      this.debug(
+        [
+          `recording event: method=${TELEMETRY_METHOD_ETLD_CHANGE}`,
+          `object=${telemetryObject} value=${telemetryValue}`,
+          `extra=${JSON.stringify(extra)}`,
+        ].join(" ")
+      );
+      browser.telemetry.recordEvent(
+        TELEMETRY_CATEGORY,
+        TELEMETRY_METHOD_ETLD_CHANGE,
+        telemetryObject,
+        telemetryValue,
+        extra
+      );
+    }
+  }
+
+  getAddonIdsForUrl(url) {
+    for (const pattern of Object.keys(this.matchPatterns)) {
+      const [urlPrefix] = pattern.split("*");
```

Comment:
Change to `.slice(0, -1)` with comment that `getMatchPatterns` returns the prefix plus `"*"`. Otherwise this logic could break if an extension uses `*` in the URL.

Reply:
if an extension uses `*` in the url, we still get a url prefix, no? (which is what I want)